src/dos/dos_code_page.cpp

```diff
@@ -0,0 +1,2223 @@
+
+	uint32_t struct_offset = 0;
+
+	uint16_t cpeh_size        = 0;
+	uint32_t next_cpeh_offset = 0;
+	uint16_t device_type      = 0;
+	std::string device_name   = {};
+	uint16_t code_page        = 0;
+	uint32_t cpih_offset      = 0;
+};
+
+struct CodePageInfoHeader {
+	static constexpr auto StructName = "CodePageInfoHeader";
+
+	uint32_t struct_offset = 0;
+
+	uint16_t version   = 0;
+	uint16_t num_fonts = 0;
+	uint16_t size      = 0;
+};
+
+struct ScreenFontHeader {
+	static constexpr auto StructName = "ScreenFontHeader";
+
+	uint32_t struct_offset = 0;
+
+	uint8_t height     = 0;
+	uint8_t width      = 0;
+	uint8_t y_aspect   = 0;
+	uint8_t x_aspect   = 0;
+	uint16_t num_chars = 0;
+};
+
+#if 0
+struct PrinterFontHeader
+{
+	// 1 - character set has to be uploaded to the printer
+	// 2 - the printer already have the character set
+	uint16_t printer_type  = 0;
+
+	// Number of bytes in escape sequences that follow; for printer type 1
+	// there are two escape sequences (first to select built-in code page,
+	// second to select the downloaded one), for printer type 2 there is
+	// only one.
+	// Escape sequences are stored in Pascal string format (first byte is
+	// a string length).
+	// Any remaining bytes up to size given in the CodePageInfoHeader are
+	// a font definition which has to be sent to the printer.
+	// Reference:
+	// - https://www.seasip.info/DOS/CPI/cpi.html
+	uint16_t escape_length = 0;
+
+	// There are no reserved bytes in the structure
+};
+#endif
+
+// DR-DOS format specific structure
+struct CharacterIndexTable {
+	static constexpr auto StructName = "CharacterIndexTable";
+
+	uint32_t struct_offset = 0;
+
+	std::vector<uint16_t> font_index = {};
+};
+
+class CpiReader {
+protected:
+	CpiReader(const std::vector<uint8_t>& content) : content(content) {}
+	virtual ~CpiReader() = default;
+
+	size_t GetContentSize() const;
+	uint32_t GetCurrentOffset() const;
+	void SetCurrentOffset(const uint32_t new_offset);
+	void SkipBytes(const uint32_t num_bytes);
+
+	bool HasUpxSignature() const;
+
+	bool IsCopyrightString() const;
+
+	std::vector<uint8_t> ReadBlob(const uint32_t length);
+
+	std::optional<FontFileHeader> ReadFontFileHeader();
+	std::optional<FontFileExtendedHeader> ReadFontFileExtendedHeader();
+	std::optional<FontInfoHeader> ReadFontInfoHeader();
+	std::optional<CodePageEntryHeader> ReadCodePageEntryHeader();
+	std::optional<CodePageInfoHeader> ReadCodePageInfoHeader();
+	std::optional<ScreenFontHeader> ReadScreenFontHeader();
+	std::optional<CharacterIndexTable> ReadCharacterIndexTable(const uint32_t num_chars);
+
+	virtual void WarnDataOutOfBounds(const uint32_t offset_start,
+	                                 const uint32_t offset_end,
+	                                 const std::string& struct_name = {}) = 0;
+	virtual void WarnStructSizeInvalid(const uint32_t offset,
+	                                   const std::string& struct_name,
+	                                   const uint32_t size) = 0;
```

Comment:
`int` or `int64` params it at all possible.

Reply:
Offsets in the CPI files are unsigned 32 bit integers (in rare cases these are pairs of 16 bit _segment:offset_ values), I would prefer to use the same type as in the file format description.

Reply:
Sure, I think this qualifies for the dealing with binary formsts case.